File: lab4.m

N = 5; %Количество регистров
len_pos = 2^N - 1; %Длина последовательности
%задаем регистры
register_x = [0, 1, 0, 1, 1];
register_y = [1, 0, 0, 1, 0];
register_x1 = [0, 1, 1, 0, 0];
register_y1 = [0, 1, 1, 0, 1];

% Генерация последовательности Голда и измененной последовательности
dis = sprintf("\nFirst Gold''s sequencen\n");
random_sequence = GoldSeq(register_x, register_y, len_pos);

dis2 = sprintf("\nSecond Gold''s sequence (x = 12, y = 13):\n");

modified_sequence = GoldSeq(register_x1, register_y1, len_pos);

%выводим последовательности на экран
for i = 1:len_pos
    dis = sprintf("%s %d", dis, random_sequence(i));
    dis2 = sprintf("%s %d", dis2, modified_sequence(i));
end
disp(dis);
disp(newline);
disp(dis2);
%копируем первую последовательность на массив для сдвига
shift_random_sequence = random_sequence;

% Находим автокорреляцию через xcorr(x,x), тк встроенная функция не работает
[cor, lag]=xcorr(random_sequence,random_sequence);
dis = sprintf("\nСдвиг");
for i = 1:len_pos
    dis = sprintf("%s|%2d", dis,i);
end
dis = sprintf("%s|корреляция|", dis);

%половина массива xcorr
t = (0:31);
corrMass = (0:31);
%считаем автокорреляцию вручную
for i = 0:len_pos
    dis = sprintf('%s\n', dis);
    dis = sprintf("%s| %2d ",dis, i);
    for j = 1:len_pos
        dis = sprintf("%s|%2d", dis, shift_random_sequence(j));
    end
    %[acf,lags] = autocorr(random_sequence);
	%Функция для рассчета автокорреляции вручную
    corr = corrXY(random_sequence, shift_random_sequence, len_pos);
    	%заполнение массива значениями корреляции
	corrMass(i+1) = corr/len_pos;
    dis = sprintf("%s| %5.2f", dis, corr/len_pos);
    %Сдвиг последовательности
    shift_random_sequence = circshift(random_sequence, [0, i+1]);
end
disp(dis);
% расчет корреляции между первой и второй последовательностями
[cor1, lag1]=xcorr(random_sequence,modified_sequence);
dis = sprintf("\nВзаимная корреляция между первоначальной и новосозданной последовательностью >> %.0f", cor1(len_pos));
disp(dis);
%Вывод графиков
subplot(2, 1, 1);
plot(t,corrMass);
xlabel("Сдвиг");
ylabel("Amplitude");
title("Ручная автокорреляция");
subplot(2, 1, 2);
plot(lag,cor);
xlabel("lag");
ylabel("Amplitude");
title("Замена автокорреляции xcorr(x,x)");

%Функция для генерации последовательности голда
function sequence = GoldSeq(register_x, register_y, len_pos)
    sequence = zeros(1, len_pos);
    for i = 1:len_pos
        sequence(i) = mod(register_x(5) + register_y(5), 2);
        [register_x, register_y] = shift(register_x, register_y);
    end
end
%Функция для рассчета автокорреляции
function corr = corrXY(random_sequence, shift_random_sequence, len_pos)
    corr = 0;
    for i = 1:len_pos
        if random_sequence(i) == shift_random_sequence(i)
            corr = corr+1;
        else 
            corr = corr-1;
        end
    end
end
%Функция для сдвига регистров
function [register_x, register_y] = shift(register_x, register_y)
    res_x = mod(register_x(3) + register_x(4), 2);
    res_y = mod(register_y(2) + register_y(3), 2);
    register_x = circshift(register_x, [0, 1]);
    register_x(1) = res_x;
    register_y = circshift(register_y, [0, 1]);
    register_y(1) = res_y;
end

File: lab4.c
#include <stdio.h>
#include <math.h>
#define N 5
#include <stdlib.h>


//сдвиг Х регистров
void reg_x(int *register_x){
	int tmp =  (register_x[2] + register_x[3]) % 2; //сумматор по модулую 2
    //Сдвиг массива
    for(int i = 0; i < N; i++){
        register_x[N-i-1] =  register_x[N-i-2];
    }
    register_x[0] = tmp; //присваиваем первому элементу суммиованное ранее значение

}
//сдвиг Y регистров
void reg_y(int *register_y){
	int tmp = (register_y[1] + register_y[2]) % 2; //сумматор по модулую 2
    //Сдвиг массива
    for(int i = 0; i < N; i++){
        register_y[N-i-1] =  register_y[N-i-2];
    }
    register_y[0] = tmp; //присваиваем первому элементу суммиованное ранее значение
}


//Получение последовательности голда
void GoldSeq(int *register_x, int *register_y, int *seq, int len_pos){
    for (int i = 0; i < len_pos; i++) {
        seq[i] = (register_x[4] + register_y[4]) % 2; //сумматор по модулю 2
        reg_x(register_x); //сдвиг регистров
        reg_y(register_y); //сдвиг регистров
    }
}

//сдвиг последовательности циклично
void shiftseQ(int *shiftSeq ,int len_pos){
    int tmp, i;
    tmp = shiftSeq[len_pos-1]; //сохранение последнего элемента
	for(i = 0;i< len_pos;i++){
        shiftSeq[len_pos-i-1] =  shiftSeq[len_pos-i-2]; //сдвиг эементов массива
    }
    shiftSeq[0] = tmp; //присваиваем первому элементу ранее сохраненное значение
}
//расчет автокорреляции
void autocorr(int *seq, int *shiftSeq, float *cor, int len_pos){
    float res = 0;
    for (int i = 0; i < len_pos; i++) {
	//считаем количество совпадений
        if (seq[i] == shiftSeq[i]){ 
            res += 1.0;
        } else{
            res -= 1.0;
        }
    }
    *cor = res;//возвращаем значения
}
//Взаимная корреляция
void crosscorr(int *seq, int *seq2, float *cor, int len_pos) {
    float res = 0.0;
    for (int i = 0; i < len_pos; i++) {
        res += seq[i] * seq2[i];
    }
    *cor=res;
}
//Выввод таблицы с информацией
void print_shift_table(int shift, int *shiftSeq, int *seq, int len_pos){
    int i,j;
    float cor;
	//Выводим первую строку таблицы
    printf("\nSdvig");
    for(i = 1; i <= len_pos; i++){
        printf("|%2d", i);
    }
    printf("|Corelation|");
	//выводим все вариации сдвига последовательности
    for(i = 0; i <= shift; i++){
        //printf("\n------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n");
	/*printf("\n");
	for(j=0;j< 174; j++){
		printf("\E(0q\E(B");
	}*/
	printf("\n");
	printf("| %2d ", i);
	//выводим значения взаимной корреляции
        for(j=0; j < len_pos; j++){
            	printf("|%2d", shiftSeq[j]);
        }
	autocorr(seq, shiftSeq, &cor, len_pos);
	printf("|%5.1f     |", cor/len_pos);
	//циклический сдвиг последовательности
	shiftseQ(shiftSeq, len_pos);
    }
    printf("\n--------------------------------------------------------------------------------------------------------------\n");
}
int main(){
    int i = 0;
    int len_pos = pow(2,N) - 1; //длина последовательности
    	//динамическое выделение памяти
	int *seq = (int*)malloc(len_pos*sizeof(int)); //Массив для хранения первой последовательности
	int *seq2 = (int*)malloc(len_pos*sizeof(int)); //Массив для хранения второй последовательности
    	int *shiftSeq  = (int*)malloc(len_pos*sizeof(int)); //Массив для храниения сдвига первой последовательности
    
	//Массивы  регистров 
	//(x = 11, y = 11+7=18)
    int register_x[N] = {0, 1, 0, 1, 1};
    int register_y[N] = {1, 0, 0, 1, 0};
    //(x = 12, y = 13)
    int register_x1[N] = {0, 1, 1, 0, 0};
    int register_y1[N] = {0, 1, 1, 0, 1};
    
	// Генерируем первую последовательность голда и выводим
    printf("\nFirst Gold's sequence :\n");
    GoldSeq(register_x, register_y, seq, len_pos);
	//копируем 1 последовательность в массив для сдвигов
	for(i = 0; i < len_pos; i++){
        	shiftSeq[i]=seq[i];
    	}

    for (i = 0; i < len_pos; i++) {
        printf("%d", seq[i]);
    }
    printf("\n\n");
	//выводим таблицу с сдвигами
    print_shift_table(len_pos, shiftSeq, seq, len_pos);

	//Генерируем вторую последовательность
    printf("\nSecond Gold's sequence :\n");
    GoldSeq(register_x1, register_y1, seq2, len_pos);
     for (i = 0; i < len_pos; i++) {
        printf("%d", seq2[i]);
    }
    printf("\n\n");
    float cor;
	//Взаимная корреляция первой и второй последовательнсти
	crosscorr(seq, seq2, &cor, len_pos);
	//printf("\nКорреляция между начальной и новосформированнной последовательностью >> %.0f\n", cor);
    printf("\nCorrelation between begin and new sequnce's >> %.0f\n", cor);
}
